### PR TITLE
Fix condition for timeout in worker task launcher

### PR DESCRIPTION
**Pull request by adarshsanjeev**
This fixes an issue where the controller would mistake a retrying worker task to timeout as soon as it is scheduled for a retry. 

When a task retries, it is possible that `recentFullyStartedWorkerTimeInMs` refers to the time when the last task for the stage started. This would cause the retrying task to timeout if the mainLoop checks the status. 

<hr>

This PR has:

- [ ] been self-reviewed.
   - [ ] using the [concurrency checklist](https://github.com/apache/druid/blob/master/dev/code-review/concurrency.md) (Remove this item if the PR doesn't have any relation to concurrency.)
- [ ] added documentation for new or modified features or behaviors.
- [ ] a release note entry in the PR description.
- [ ] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [ ] added or updated version, license, or notice information in [licenses.yaml](https://github.com/apache/druid/blob/master/dev/license.md)
- [ ] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [ ] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [ ] added integration tests.
- [ ] been tested in a test Druid cluster.

### CODE REVIEW

File: extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/indexing/MSQWorkerTaskLauncher.java [hunk 1]
```diff
@@ -762,9 +762,11 @@ public boolean didFail()
      */
```
Comment (LakshSingla) [inline extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/indexing/MSQWorkerTaskLauncher.java:762]: Can you please update the Javadoc for this method? Also please mention both the conditions separately and what they guardrail, i.e. `currentTimeMillis - recentFullyStartedWorkerTimeInMs.get() > maxTaskStartDelayMillis` ensures that no new worker has been launched in the last maxTaskStartDelayMillis.

**Review by LakshSingla — COMMENTED** [review]
There were a few comments that I left on the PR https://github.com/apache/druid/pull/14172, can you please address them along this review as well? They are mostly stylistic changes.

Also, can we test this class somehow? The way it is written, it seems difficult given that most of the timeouts are hardcoded, and we won't want to wait 10mins for a unit test to fail/pass, but perhaps we can do it by manipulating the private variables using reflection, or refactor the class?

File: extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/indexing/MSQWorkerTaskLauncher.java (hunk 1 above)
Comment (adarshsanjeev) [reply extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/indexing/MSQWorkerTaskLauncher.java:767]: Done

File: extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/indexing/error/TaskStartTimeoutFault.java [hunk 2]
```diff
@@ -32,13 +32,13 @@ public class TaskStartTimeoutFault extends BaseMSQFault
...
 
-  private final int numTasksNotStarted;
+  private final int pendingTasks;
   private final int totalTasks;
   private final long timeout;
 
   @JsonCreator
   public TaskStartTimeoutFault(
-      @JsonProperty("numTasksNotStarted") int numTasksNotStarted,
+      @JsonProperty("pendingTasks") int pendingTasks,
```
Comment (LakshSingla) [inline extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/indexing/error/TaskStartTimeoutFault.java:41]: This warrants a change in the docs at `druid/docs/multi-stage-query/reference.md`.

**Review by LakshSingla — APPROVED** [review]
LGTM apart from doc changes.

File: extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/indexing/error/TaskStartTimeoutFault.java (hunk 2 above)
Comment (adarshsanjeev) [reply extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/indexing/error/TaskStartTimeoutFault.java:41]: Thanks! Changed.

**Comment by cryptoe** [comment]
@adarshsanjeev Nice catch. Apologies for introducing this bug.

**Pull request merged**